font/src/directwrite/mod.rs

```diff
@@ -13,33 +13,207 @@
     }
 
-    fn metrics(&self, key: FontKey, size: Size) -> Result<Metrics, Error> {
-        let font = self.fonts.get(key.token as usize).ok_or(Error::FontNotLoaded)?;
+    fn rasterize_glyph(&mut self, glyph: GlyphKey) -> Result<RasterizedGlyph, Error> {
+        let font = &self.fonts.get(&glyph.font_key).ok_or(Error::FontNotLoaded)?.font_face;
+
+        let offset = GlyphOffset { advanceOffset: 0.0, ascenderOffset: 0.0 };
+
+        let glyph_index = *font
+            .get_glyph_indices(&[glyph.c as u32])
+            .first()
+            .ok_or_else(|| Error::MissingGlyph(glyph.c))?;
+
+        // The DirectWrite documentation states that we should get 0 returned if the glyph
+        // does not exist in the font
+        if glyph_index == 0 {
+            return Err(Error::MissingGlyph(glyph.c));
+        }
+
+        let em_size = em_size(glyph.size);
+
+        let glyph_run = dwrote::DWRITE_GLYPH_RUN {
+            fontFace: unsafe { font.as_ptr() },
+            fontEmSize: em_size,
+            glyphCount: 1,
+            glyphIndices: &(glyph_index),
+            glyphAdvances: &(0.0),
+            glyphOffsets: &(offset),
+            isSideways: 0,
+            bidiLevel: 0,
+        };
+
+        let rendering_mode = font.get_recommended_rendering_mode_default_params(
+            em_size,
+            self.device_pixel_ratio,
+            dwrote::DWRITE_MEASURING_MODE_NATURAL,
+        );
+
+        let glyph_analysis = GlyphRunAnalysis::create(
+            &glyph_run,
+            self.device_pixel_ratio,
+            None,
+            rendering_mode,
+            dwrote::DWRITE_MEASURING_MODE_NATURAL,
+            0.0,
+            0.0,
+        )
+        .or_else(|_| Err(Error::MissingGlyph(glyph.c)))?;
+
+        let bounds = glyph_analysis
+            .get_alpha_texture_bounds(dwrote::DWRITE_TEXTURE_CLEARTYPE_3x1)
+            .or_else(|_| Err(Error::MissingGlyph(glyph.c)))?;
+
+        let buf = glyph_analysis
+            .create_alpha_texture(dwrote::DWRITE_TEXTURE_CLEARTYPE_3x1, bounds)
+            .or_else(|_| Err(Error::MissingGlyph(glyph.c)))?;
+
+        Ok(RasterizedGlyph {
+            c: glyph.c,
+            width: (bounds.right - bounds.left) as i32,
+            height: (bounds.bottom - bounds.top) as i32,
+            top: -bounds.top,
+            left: bounds.left,
+            buf: BitmapBuffer::RGB(buf),
+        })
+    }
+
+    fn rasterize_fallback_glyph(
+        &mut self,
+        glyph: GlyphKey,
+        original_glyph: GlyphKey,
+    ) -> Result<RasterizedGlyph, Error> {
+        // The fallback font is unlikely to have the same advance size as the monospace font.
+        // Therefore we may need to dynamically decrease the font size so the glyph fits in the
+        // cell width.
+        //
+        // We'll only do this for any glyph > 2 cells wide, downsizing it to be just 2 cells wide.
+        // This should help keep things tidy without resizing many glyphs needlessly.
```

Comment:
Isn't a halfwidth font just as likely to exceed the width as a fullwidth one?

It seems a bit strange that the widths would differentiate so strongly, since our fonts sizes are specified in points.

Reply:
> Isn't a halfwidth font just as likely to exceed the width as a fullwidth one?

Yes, a halfwidth glyph might overflow one cell just as a fullwidth glyph might overflow two cells.

Now I've removed the `unicode-width` dependency I don't see a way to reliably handle the halfwidth case. (I cannot find any DirectWrite API to ask if a codepoint is half- or full-width.)

Reply:
> Is there some documentation on this being required, or recommended? Or did you mainly look at the source code of the windows terminal without a proper name?

There's no documentation I've found on it. I think it's just a sad fact that Windows' emoji font has completely different widths to my monospace font. Despite Windows' having selected this font supposedly as a suitable fallback!

This algorithm is what I ended up with after a few varied attempts:

- My initial implementation had no scaling, and I saw glyphs with slight offset issues like in [this comment](https://github.com/alacritty/alacritty/pull/3127#issuecomment-569404916)
- To fix the offset problem I then attempted downscaling [like in the OP](https://github.com/alacritty/alacritty/pull/3127#issue-357228997). TBH these images don't look great. At this point I was already using glyph advance to calculate the factor to downscale by.
- I then decided to peek at what the Windows Terminal was doing. I felt their strategy of scaling the font size could work here too, which is why I then decided to implement the same.

It's not great, but I think it looks better than either the out-of-bounds fonts or the downsampled ones from the OP.

Reply:
I personally find it questionable to implement such a thing on Windows, while opting to overlap on other platforms. We should be consistent across platforms.

And as we found out in this PR, the process of downscaling isn't simple either. Scaling the output glyph can easily lead to poor looking fonts so it's not really an option. But scaling the font size can not guarantee that the result actually works properly, especially with fonts that are not strictly monospace (which is likely the case for emojis).

Reply:
>Neither macOS nor Linux/BSD require scaling fonts after fallback. It's strange to me that this would be required on Windows, without providing a recommended way of dealing with this.

Speaking of Linux here, it does require your app to scale things(fixup_factor) and it can also provide an odd scale factor, since user can just set one in a config. So technically, we're scaling on Linux to a user defined point, so it's just smarter than whatever Windows is providing.

>I personally find it questionable to implement such a thing on Windows, while opting to overlap on other platforms. We should be consistent across platforms.

We will be consistent with such change on all platforms, since our overlapping is related to `wcwidth`, however this PR doesn't speculate it and just kind of using more smarter `downsampling` than for example Linux/BSD impl.

>And as we found out in this PR, the process of downscaling isn't simple either. Scaling the output glyph can easily lead to poor looking fonts so it's not really an option. But scaling the font size can not guarantee that the result actually works properly, especially with fonts that are not strictly monospace (which is likely the case for emojis).

They are full-width fonts, we can perform such scaling only on colored fonts tbh, since we can query such information IIRC.

Anyway, @davidhewitt, could you show me how this string looks like `❄m❄` on your PR?

Reply:
> Anyway, @davidhewitt, could you show me how this string looks like `❄m❄` on your PR?

Sure, it looks like this when pasted in (I typed a few spaces after to get the cursor out the way):

![image](https://user-images.githubusercontent.com/1939362/72479306-068dfe80-37ec-11ea-9d4c-ac08a93c2a08.png)

Reply:
One more note, some odd fonts on Linux like `Inconsolata` are playing really bad with all the other fonts on Linux, so technically even normal chars from other fonts are starts to overlap each other, not speaking of emojis. So technically @chrisduerr is right here, I guess.

Reply:
Ok. It looks like the consensus is to just not do the scaling, so I'll remove it now.

Reply:
>Sure, it looks like this when pasted in (I typed a few spaces after to get the cursor out the way):

The picture is absolutely correct though, we and the most of other terminals, have the same thing on Linux. So now it's just a matter of using the same. However we can implement the same algo for font change on Linux, it should be pretty trivial.

Reply:
@davidhewitt wait a bit.

Reply:
Ok. I'm going to head to bed for the night; will pick this up at a later date when a decision has been made. My personal preference is to have the scaling in so the emojis line up nicely, but I will remove it if that's the consensus from everyone else.

We have some time to mull this over as we need the `dwrote` PRs merged before this is mergable now anyway.

Reply:
After some discussion on IRC, and further investigation with our Linux/BSD backend, we decided that we shouldn't scale things for now and hold it back until there will be a lot of complains in the future. The algorithm since a bit of fragile tbh, since some font metrics are not that much of reliable, knowing that we can fallback to a glyph from non monospace font and there's always tons of use cases.

Also we do have a font rfc #957 and there's an idea on providing some set of options to a font, so with it you can just change a font size to "bad" fonts in your fallback to better match a primary font, since it's more reliable way to do things than guessing based off some font metrics.

tl;dr You issue should be covered by #957 and we should revert any scaling to match our other backends for now.